backends/smt2/smtbmc.py

```diff
@@ -969,6 +975,38 @@ def write_vcd_trace(steps_start, steps_stop, index):
 
         vcd.set_time(steps_stop)
 
+def find_state_loop(steps_start, steps_stop):
+    print_msg(f"Looking for loops")
+
+    path_list = list()
+
+    for netpath in sorted(smt.hiernets(topmod)):
+        hidden_net = False
+        for n in netpath:
+            if n.startswith("$"):
+                hidden_net = True
+        if not hidden_net:
+            path_list.append(netpath)
+
+    mem_trace_data = collect_mem_trace_data(steps_start, steps_stop)
+
+    states = dict()
+    path_tupled_list = [tuple(p) for p in path_list]
+    for i in range(steps_start, steps_stop):
+        value_list = smt.get_net_bin_list(topmod, path_list, "s%d" % i)
+        state = tuple(zip(path_tupled_list, value_list))
+        if states.get(state):
+            return (i, states[state])
+        else:
+            states[state] = i
+
+        if i in mem_trace_data:
```

Comment:
This check for bailing out has to go above the loop-detection part, otherwise you will get false positives when a design with memory repeats the same non-memory state with changed memory contents.

Also I think using
```python
mem_state = sorted(
    (tuple(path), addr, data) for path, addr, data in words)
    for words in mem_trace_data.get(i, [])
)
state = tuple(value_list), tuple(mem_state)
```
above might be all that's needed to also support memory. I haven't tested that at all yet. Even if this works it's ok to add memory support later.

Reply:
Implemented fully, tested on modified examples to a limited extent